### PR TITLE
12.0 base user role profile

### PR DESCRIPTION
**Pull request by kevinkhao**
add profile functionality to roles:
- profiles to decide which roles the user wants to use at the time
- new widget to choose user profile, similar to the widget to choose company in a multicompany environment

please check https://github.com/OCA/server-backend/pull/68 for the commits not related to profile

### CODE REVIEW

**Review by jcdrubay — APPROVED** [review]

File: base_user_role/models/user.py [hunk 1]
```diff
@@ -49,6 +54,14 @@ def write(self, vals):
         self.sudo().set_groups_from_roles()
         return res
 
+    def _get_enabled_roles(self):
```
Comment (jcdrubay) [inline base_user_role/models/user.py]: Could be renamed to `_get_applicable_roles` (something different from get_enabled_roles) to highlight that it's filtering on more than the field `is_enabled`.

**Comment by jcdrubay** [comment]
Could you add in the description that we should start by reviewing:
https://github.com/OCA/server-backend/pull/68#pullrequestreview-369985033

**Comment by kevinkhao** [comment]
This module is abstract, so an example module has been added, separated in a new module that depends on a lot of different modules to demo functionality for different permissions/groups

**Review by sebastienbeau — APPROVED** [review]
LGTM (code review + test)
Thanks for this usefull module !
This will avoid to the user to mess with config when their are working

File: base_user_role_profile/models/user.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+            "params": {"menu_id": menu.id},
+        }
+
+    def action_profile_change(self, vals):
+        self.write(vals)
+        return self._get_action_root_menu()
+
+    @api.model
+    def create(self, vals):
+        new_record = super(ResUsers, self).create(vals)
```
Comment (bealdav) [inline base_user_role_profile/models/user.py]:
```suggestion
        new_record = super().create(vals)
```

File: base_user_role_profile/models/user.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        if vals.get("company_id") or vals.get("role_line_ids"):
+            new_record.sudo()._compute_profile_ids()
+        return new_record
+
+    def write(self, vals):
+        # inspired by base/models/res_users.py l. 491
+        if self == self.env.user and vals.get("profile_id"):
+            self.sudo().write({"profile_id": vals["profile_id"]})
+            del vals["profile_id"]
+        res = super(ResUsers, self).write(vals)
```
Comment (bealdav) [inline base_user_role_profile/models/user.py]:
```suggestion
        res = super().write(vals)
```

**Review by bealdav — APPROVED** [review]
Except typo OK for me

**Comment by sebastienbeau** [comment]
@bealdav typo have been fixed and PR rebased

**Pull request merged**